Add region annotation feature to 2D and 3D heatmaps

zenWai opened the pull request:
**What is this PR**

- [X] Addition of a new feature

**Why is this PR needed?**

When visualizing heatmaps, it's crucial for researchers to be able to identify specific regions directly on the visualization. Other use cases might apply.

Currently, users need to:

- Manually cross-reference regions with external references
- Rely on prior knowledge of brain anatomy to interpret

## New Features

- **Optional `annotate_regions` parameter** supports three modes:

1. `(bool)`: Enable/disable annotation for all regions.
2. `(List[str])`: Annotate specific regions.
3. `(Dict[str, Union[str, int, float]])`: Annotate regions with custom text or values.

- **Optional ``annotate_text_options_2d``**:

`(Dict)`: supports matplotlib.text arguments

- **Helper Function ``find_annotation_position_inside_polygon``**:

Helper to find the best position for text annotations within a 2D polygon.

- **Helper Function ``get_region_annotation_text``**:

Helper to handle the annotation text for a region if annotation is expected.

## Example

<table>
  <tr>
    <td width="30%"><img src="https://github.com/user-attachments/assets/69d10e7f-88eb-4ccf-8044-de31b470c6b1" width="100%"></td>
    <td width="30%"><img src="https://github.com/user-attachments/assets/0395a3e0-2435-4ea8-afe6-93fa555cfdc3" width="100%"></td>
    <td width="40%"><img src="https://github.com/user-attachments/assets/93d64683-8286-4878-934e-d9e1d9747ce3" width="100%"></td>
  </tr>
</table>

## References

Please reference any existing issues/PRs that relate to this PR.

## How has this PR been tested?

Locally with /examples/ and personal data.

## Is this a breaking change?

No.

## Does this PR require an update to the documentation?

Does not require.

## Checklist:

- [X] The code has been tested locally
- [X] Tests have been added to cover all new functionality (unit & integration)
- [ ] The documentation has been updated to reflect any changes
- [X] The code has been formatted with [pre-commit](https://pre-commit.com/)

alessandrofelder commented on brainglobe_heatmap/heatmaps.py:
```diff
@@ -48,6 +50,187 @@ def check_values(values: dict, atlas: Atlas) -> Tuple[float, float]:
+        """
+        point = np.array([px, py])
+
+        if not polygon.contains_point((px, py)):
```
This seems to work, but wouldn't it be better practice to pass `polygon` as an argument here to the function (it technically doesn't get defined until further down?

zenWai replied:
_The polygon is a matplotlib.path.Path object. The reason for using it is for easy access to contains_points and contains_point methods, it's the approach I found to avoid major performance issues._

Well spotted! It is defined outside the scope of calculate_point_to_edges_distance. I will take a look at this.

alessandrofelder commented on brainglobe_heatmap/heatmaps.py:
```diff
@@ -48,6 +50,187 @@ def check_values(values: dict, atlas: Atlas) -> Tuple[float, float]:
+    polygon = pltpath.Path(polygon_vertices)
+
+    # determine the bounding box of the polygon
+    minx, miny = np.min(polygon_vertices, axis=0)
+    maxx, maxy = np.max(polygon_vertices, axis=0)
```
```suggestion
    minx = np.min(polygon_vertices, axis=0)
    miny = np.min(polygon_vertices, axis=1)
```
I would naively think?
Is this currently a bug, or do I need an explanation of the `pltpath.Path` stores polygonal vertices? :smile:

zenWai replied:
It is working as intended, looking at this snippet I reckon it would be a good idea to change variable naming to min_x min_y and max_x min_y to be more pythonic. I will take a look.

I will try to explain,
_I am forcing polygon_vertices to numpy.ndarray(np.ndarray)._
When you do, 
- np.min(polygon_vertices, axis=0) will give the minimum value from each column, good for our needs, one array with [minx, miny].
- When you do, np.min(polygon_vertices, axis=1) will give the minimum value within each row, one array with one value per row.

```python
test = [
   [1, 2],
   [3, 4],
   [5, 6],
   [6, 1]
]
array_with_min_x_and_miny = np.min(test, axis=0) # [1 1]
array_with_min_values_of_each_vertice = np.min(test, axis=1) # [1 3 5 1]
```

alessandrofelder replied:
I see, all clear - thanks for the explanation!

alessandrofelder reviewed (commented):
Thanks for the PR @zenWai - I think the added functionality is awesome and useful, and seems to work as expected for the two examples I tried locally. The docstrings you've added are great.

I'm not sure you think it's ready for review yet (let me know) -  I've had an initial look and some tiny comments.

zenWai commented:
- changed import matplotlib.path as pltpath to as mpath
- changed`minx, miny, maxx, maxy` to `min_x, min_y, max_x, max_y`
- moved check `polygon.contains_point((px, py))`

zenWai commented:
### Regarding the `find_annotation_position_inside_polygon` function

I'd like to discuss `find_annotation_position_inside_polygon` function.
There is a well-known and thoroughly tested package called [shapely](https://github.com/shapely/shapely) that has a built-in function performing the same task. By introducing shapely as a dependency, we could replace the custom function with this:

```python
# returns the optimal position for placing an annotation inside a polygon.
polygon = Polygon(coords.tolist()) # create a shapely polygon object
label = polylabel(polygon, tolerance=0.1) # label.x label.y
```

I prepared a snippet for a quick testing on the implemented functionality vs the shapely functionality,
```python
            # After - name, segment = r.split("_segment_")
            import time
            from shapely import Polygon
            from shapely.algorithms.polylabel import polylabel

            start_time = time.perf_counter()
            polygon = Polygon(coords.tolist())
            label = polylabel(polygon, tolerance=0.1)
            polylabel_time = time.perf_counter() - start_time
            print(name)
            print("Polylabel result:", label)
            print("Polylabel execution time: {:.4f} seconds".format(polylabel_time))

            start_time = time.perf_counter()
            x, y = find_annotation_position_inside_polygon(coords, precision=0.1)
            find_annotation_time = time.perf_counter() - start_time
            print("find_annotation_position result:", x, y)
            print("find_annotation_position execution time: {:.4f} seconds".format(find_annotation_time))
```

The behaviour is similar from my testings,
_The function `find_annotation_position_inside_polygon` exists primarily to avoid introducing an additional dependency._

I'm interested in your thoughts on whether maintaining `find_annotation_position_inside_polygon` is preferable to adding shapely as a dependency.
In my opinion `find_annotation_position_inside_polygon` does the job well 👌

alessandrofelder reviewed (approved):
Thanks @zenWai. I unfortunately don't have the capacity to review the code beyond
* I can see that the code does some sensible-looking geometric operations ✅ 
* I kind of understand the algorithm on a conceptual level ✅ 
* the examples work as expected ✅ 
* the feature is definitely well-motivated and useful ✅ 

>I'm interested in your thoughts on whether maintaining find_annotation_position_inside_polygon is preferable to adding shapely as a dependency.

You've already pre-empted the major pro and con, so I don't have much to add. I can discuss this at our developer meeting tomorrow (you are welcome to [join if you can/like](https://brainglobe.zulipchat.com/#narrow/channel/414089-developer-meeting/topic/developer-meeting-20-02-2025) - needs sign-up, you can use your GH account).

FWIW I am happy to merge as is now (and fall back on `shapely` if we find bugs we can't fix in the current code) - unless someone convinces me otherwise tomorrow, I will do so.

adamltyson commented:
Thanks for this @zenWai, I think this will be useful. Before merging though, a few things come to mind:
- Is it possible to add similar functionality for 3D heatmaps too? Ideally, there should be as little difference between 2D and 3D, otherwise it will make it less intuitive for users
- IMO the additions to the examples make them harder to understand as basic "get started" scripts. Would you mind reverting all the changs to the examples, and adding a single `region_annotation.py` example that best showcases these new changes?
- Our online docs for brainglobe-heatmap are admittedly not great, but would you mind adding something to the docs [here](https://github.com/brainglobe/brainglobe.github.io/tree/main/docs/source/documentation/brainglobe-heatmap). If you're not sure where it should best go, we can move it around. 
- We really need to start enforcing test coverage on this repo as we do with other BrainGlobe tools. Do you think there are any parts of this PR that could be tested in isolation?

zenWai commented:
> Thanks for this @zenWai, I think this will be useful. Before merging though, a few things come to mind:
> 
> * Is it possible to add similar functionality for 3D heatmaps too? Ideally, there should be as little difference between 2D and 3D, otherwise it will make it less intuitive for users

✅

> * IMO the additions to the examples make them harder to understand as basic "get started" scripts. Would you mind reverting all the changs to the examples, and adding a single `region_annotation.py` example that best showcases these new changes?

✅

> * Our online docs for brainglobe-heatmap are admittedly not great, but would you mind adding something to the docs [here](https://github.com/brainglobe/brainglobe.github.io/tree/main/docs/source/documentation/brainglobe-heatmap). If you're not sure where it should best go, we can move it around.

I will later for sure.

> * We really need to start enforcing test coverage on this repo as we do with other BrainGlobe tools. Do you think there are any parts of this PR that could be tested in isolation?

I added test for functions `find_annotation_position_inside_polygon` and `get_region_annotation_text`. Did int test for the example.

Thank you so much @alessandrofelder and @adamltyson for taking your take to check this PR.
I need some guidance on the tests and if this is what you were expecting.
Looking forward for your insight.

alessandrofelder reviewed (commented):
Well done on adding and fixing the tests @zenWai - we just went from 0 to 70% test coverage :tada: :rocket: :smile: 

An update to our docs would be much appreciated.

I have some minor nitpicky comments, and a major request around how the labels are displayed in 3D (I don't know how easy this is!) - let us know whether you would be available to help with it (no worries if not!)

zenWai commented on brainglobe_heatmap/heatmaps.py:
```diff
@@ -110,6 +144,17 @@ def __init__(
+            If a list, annotates only the specified regions.
+            If a dict, uses custom text/values for annotations.
+            Default is False.
+        annotate_text_options_2d : dict, optional
+            Options for customizing region annotations text in 2D format.
+            matplotlib.text parameters
+            Default is None
```
Currently only for 2D, supports any [matplotlib.text parameters ](https://matplotlib.org/stable/api/text_api.html#matplotlib.text.Text) and pass the parameters to the ax.annotate on plot_subplot(). The example  have some of the text options I think are most useful.
Currently 3d is implemented using the make_actor_label from brainrender, self.scene.add_label() function.
For 3d there is the option to pass this:
```
- size: int, text size. Default 300
- color: str, text color. A list of colors can be passed
             if None a gray color is used. Default None.
- xoffset, yoffset, zoffset: integers that shift the label position
- radius: radius of sphere used to denote label anchor. Set to 0 or None to hide.
```
We could revisit this after we [figure out the best way of doing the 3D](https://github.com/brainglobe/brainglobe-heatmap/pull/77#discussion_r1977907338)

zenWai commented on examples/region_annotation.py:
```diff
@@ -0,0 +1,54 @@
+    alpha=1,  # float in range 0-1
+)
+# ----------------------------------------
+# Annotate specified regions,
+# annotate_regions = ["TH", "MOs"]
+#
+# Annotate regions with custom values or text,
+# annotate_regions = dict(MO=0.4, SS="Text")
+
+f = bgh.Heatmap(
+    values,
+    # when using a named orientation, you can pass a single value!
+    position=9000,
+    # 'frontal' or 'sagittal,' or 'horizontal' or a tuple (x,y,z)
+    orientation="frontal",
+    thickness=3000,
+    title="frontal view",
+    vmin=-5,
+    vmax=3,
+    annotate_regions=annotate_regions,
+    annotate_text_options_2d=annotate_text_options_2d,
+    format="2D",  # or 3D
+).show()
```
I understand for sure, let's revisit the examples and texts and talk about the best approach in detail after [figure out the best way of doing the 3D](https://github.com/brainglobe/brainglobe-heatmap/pull/77#discussion_r1977907338) sounds good?

zenWai commented on tests/test_integration/test_example_region_annotation.py:
```diff
@@ -0,0 +1,199 @@
+        ([], [], EXPECTED_VISIBLE_REGIONS + EXPECTED_ABSENT_REGIONS),
+        ({}, [], EXPECTED_VISIBLE_REGIONS + EXPECTED_ABSENT_REGIONS),
+        # specified regions
+        (["TH", "SS"], ["TH"], ["SS"] + EXPECTED_ABSENT_REGIONS),
```
Yes, the approach of the test example follows hard-coded expected regions and hard-coded expected absent regions.
"SS" is redundant and already present on the EXPECTED_ABSENT_REGIONS.
The idea to have it there was to assert that, while we choose to annotate TH and SS(w/ list annotate only the given regions), matplotlib.axes.Axes.annotate is not called.
I did the same on # Custom annotations.
I am open to talk about a different approach for the example test or tweaks

zenWai commented on examples/region_annotation.py:
```diff
@@ -0,0 +1,54 @@
+"""
+This example shows how generate a heatmap
+in 2D and 3D with annotations on brain regions
```
For sure it is not ideal, I think a separate PR to brainrender on function make_actor_label() might do the trick, I did try some things, I was getting unexpected behaviours and can't see how to change it successfully.
I will explain visually the current behaviour by using only CB brain region,

| Image 1 | Image 2 | Image 3 |
|---------|---------|---------|
| ![left](https://github.com/user-attachments/assets/5185f5b4-cf33-4769-9333-df3daf48a8c0) | ![middle](https://github.com/user-attachments/assets/065124d3-4640-4373-a5db-842412a13feb) | ![right](https://github.com/user-attachments/assets/4279c240-f9cf-4781-80c2-22b9ad87e579) |

From my understanding,
_Sphere and text_ find the most top position of the brain.
The _Sphere is sucessfully_ finding a good position in the region that is between plane0 and plane1 that is defined by the position and thickness.
The _Text is unsucessfully_ finding a good position, it is stationary and looks like a optimal position if no slicing was being executed, i.g full actor VS sliced actor.

- I did try to play arround with make_actor_label() unsucessfully.
- I did try to make a custom annotation function for heatmap unsucessfully.

The perfect approach would be to align the text with the slice0, find a good position on the actor, and replicate what 2d annotation is doing, and annotate every actor that touches the plane0?
- I did try that, unsucessfully hmmm.

I did come up with a [solution](https://github.com/brainglobe/brainglobe-heatmap/pull/77/commits/b0412ac8c69100a18fd6b99025837bc6fa226fda), that fix the text is coming out of the viewing plane, with issues on sagittal, I will post some examples,
| Image 1 | Image 2 | Image 3 |
|---------|---------|---------|
| ![left_2](https://github.com/user-attachments/assets/5662b494-e99c-496d-8cd6-3adf3007b10f) |![middle_2](https://github.com/user-attachments/assets/542cd1aa-c088-4a56-a0cf-e7d961480718) | ![right_2](https://github.com/user-attachments/assets/7678675b-6a91-496e-baa9-9a8a9532dbbc) |
This approach will place the sphere and text near the slice plane0 and will annotate any region that touches the plane0 vertices.
| ex. example.py | ex. horizontal 1 | ex. horizontal 2 |
|---------|---------|---------|
| ![example](https://github.com/user-attachments/assets/05362531-01ca-4bff-b69b-4ba242ccb45b) | ![horizontal_1](https://github.com/user-attachments/assets/5d03a04b-9a76-4fcc-ac8c-8bbbdd1c07b1) | ![horizontal_2](https://github.com/user-attachments/assets/aeb4b0ba-a86c-4565-b49d-2bedba5b6699) |

Major Issue: The function ```self.slicer.plane0.intersect_with()```  used for this implementation always gives vertices <0 when orientation is 'sagittal' and I don't understand why.

Open to suggestions.

alessandrofelder replied:
All good - thanks for the detailed descriptions. Agree we need to fix that in `brainrender`'s `add_label`

zenWai commented on brainglobe_heatmap/heatmaps.py:
```diff
@@ -48,6 +51,35 @@ def check_values(values: dict, atlas: Atlas) -> Tuple[float, float]:
+    if polygon.geom_type == "MultiPolygon" and isinstance(
+        polygon, MultiPolygon
+    ):
+        polygon = max(polygon.geoms, key=lambda p: p.area)
```
[Done](https://github.com/brainglobe/brainglobe-heatmap/pull/77/commits/744dea9381ddab3cd3594a381ed225d06f6714da)

alessandrofelder commented:
Thanks for this @zenWai - just for info, I probably won't find time to review this until end of the week

alessandrofelder reviewed (commented):
Cool! Thanks a lot @zenWai - I am happy to merge this once we have 3 examples (each showing a different way of annotating, and one of them 3D). We can solve `add_label` orientation on the `brainrender` side: you are welcome to tackle that if you have time (otherwise one-day the core team will make the time :) ).

zenWai commented:
Thank you for all the time you have been taking on this PR.

> We can solve `add_label` orientation on the `brainrender` side: you are welcome to tackle that if you have time (otherwise one-day the core team will make the time :) ).
I don't think I am capable of it.

In that case the 3D implementation will go back to the simple,
```python
                self.scene.add_label(
                    actor=region_actor,
                    label=display_text,
                )
```
And then be handled on brainrender side?

zenWai commented:
- Reverted the possible solution for 3d.
- Split example into 3

| `region_annotation.py` | `region_annotation_custom.py` | `region_annotation_specified.py` |
|:-----------------:|:------------------:|:----------------:|
| ![region_annotation](https://github.com/user-attachments/assets/1e0dc0c8-9f23-4ec1-9914-1e6de791700a) | ![region_annotation_custom](https://github.com/user-attachments/assets/4e01d2c8-0bd5-4dee-8b55-d3be9d383b45) | ![region_annotation_specified](https://github.com/user-attachments/assets/697c1a33-930e-4d59-8aac-13f47700581e) |
| 2D w/ Boolean annotate_regions=True | 3D w/ Dict annotate_regions=dict(region=custom_ann) | 2D w/ List annotate_regions=["region"] |

issue on brainrender is open to improve/fix visibility of 3d annotations on sliced actors https://github.com/brainglobe/brainrender/issues/415

@alessandrofelder Feel free to ask for changes 👌

alessandrofelder reviewed (approved):
Awesome - happy with this.
Confirming I was able to successfully run new examples locally :heavy_check_mark: 
Thanks a lot @zenWai
